odoo/api.py

```diff
@@ -34,12 +33,31 @@
 from .tools.translate import _
 from odoo.tools.misc import StackMap
 
-if TYPE_CHECKING:
+import typing
+if typing.TYPE_CHECKING:
+    from collections.abc import Callable
     from odoo.sql_db import Cursor, TestCursor
     from odoo.models import BaseModel
+    from .models import NewId
+    try:
+        from typing_extensions import Self  # noqa: F401
+    except ImportError:
+        from typing import Self  # noqa: F401
+    M = typing.TypeVar("M", bound=BaseModel)
+    # type written as union because not recognized by mypy when imported
+    IdType = typing.Union[int, NewId]  # noqa: UP007
+else:
+    Self = None
+    M = typing.TypeVar("M")
+    IdType = int
```

Comment:
NewId is not available in running time, can be only use for typing

Reply:
That's a recipe for bugs, because such a definition is **incorrect** at runtime!  Someone doing `isinstance(x, IdType)` might get a different result than expected.  And you cannot say: "Yeah, it is available, but don't use it." :wink:

Reply:
Indeed, it should only be available for typing or it should be correct.
However, in normal execution, there is a lot of cycles between models.py, fields.py, api.py. Solving that cycle seems quite difficult.

Reply:
```suggestion
    IdType: typing.TypeAlias = int | NewId
```
Also, solve the cyclic dependency issue by moving the definition of class `NewId` to api.py.